programs/verified_stake/src/lib.rs

```diff
@@ -0,0 +1,585 @@
+/*
+ * This file is part of Albus code.
+ *
+ * Copyright (c) 2023, mFactory GmbH
+ *
+ * Albus is free software: you can redistribute it
+ * and/or modify it under the terms of the GNU Affero General Public License
+ * as published by the Free Software Foundation, either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * Albus is distributed in the hope that it
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.
+ * If not, see <https://www.gnu.org/licenses/agpl-3.0.html>.
+ *
+ * You can be released from the requirements of the Affero GNU General Public License
+ * by purchasing a commercial license. The purchase of such a license is
+ * mandatory as soon as you develop commercial activities using the
+ * Albus code without disclosing the source code of
+ * your own applications.
+ *
+ * The developer of this program can be contacted at <info@albus.finance>.
+ */
+
+use anchor_lang::{prelude::*, solana_program::stake};
+
+declare_id!("CMev81L3acPrcTTevCFGdcNQnDypMGzuiAUgo8NBZJzr");
+
+#[program]
+pub mod verified_stake {
+    use albus_verifier::check_compliant;
+    use anchor_lang::{
+        prelude::borsh::{BorshDeserialize, BorshSerialize},
+        solana_program::program::invoke,
+    };
+
+    use super::*;
+
+    #[derive(BorshSerialize, BorshDeserialize)]
+    pub enum StakeAuthorize {
+        Staker,
+        Withdrawer,
+    }
+
+    impl Into<stake::state::StakeAuthorize> for StakeAuthorize {
+        fn into(self) -> stake::state::StakeAuthorize {
+            match self {
+                StakeAuthorize::Staker => stake::state::StakeAuthorize::Staker,
+                StakeAuthorize::Withdrawer => stake::state::StakeAuthorize::Withdrawer,
+            }
+        }
+    }
+
+    pub fn split(ctx: Context<VerifiedSplit>, lamports: u64) -> Result<()> {
+        check_compliant(
+            &ctx.accounts.zkp_request,
+            Some(ctx.accounts.authorized.key()),
+        )?;
+
+        let ixs = stake::instruction::split(
+            &ctx.accounts.stake.key(),
+            &ctx.accounts.authorized.key(),
+            lamports,
+            &ctx.accounts.split_stake.key(),
+        );
+
+        // determine that the number of instructions is equal to 3
+        assert_eq!(ixs.len(), 3);
```

Comment:
Don't use asserts in production code... You can replace it with `debug_assert_eq`